monai/transforms/meta_matrix.py

```diff
@@ -0,0 +1,292 @@
+        matrix = _create_rotate_90(self._dims, rotations, axis)
+        return MetaMatrix(matrix, extra_args)
+
+    def flip(self, axis, **extra_args):
+        matrix = _create_flip(self._dims, axis, self._eye)
+        return MetaMatrix(matrix, extra_args)
+
+    def shear(self, coefs: Union[Sequence[float], float], **extra_args):
+        matrix = _create_shear(self._dims, coefs, self._eye)
+        return MetaMatrix(matrix, extra_args)
+
+    def scale(self, factors: Union[Sequence[float], float], **extra_args):
+        matrix = _create_scale(self._dims, factors, self._diag)
+        return MetaMatrix(matrix, extra_args)
+
+    def translate(self, offsets: Union[Sequence[float], float], **extra_args):
+        matrix = _create_translate(self._dims, offsets, self._eye)
+        return MetaMatrix(matrix, extra_args)
+
+
+def ensure_tensor(data: NdarrayOrTensor):
+    if isinstance(data, torch.Tensor):
+        return data
+
+    return torch.as_tensor(data)
+
+
+def apply_align_corners(matrix, spatial_size, factory):
+    inflated_spatial_size = tuple(s + 1 for s in spatial_size)
+    scale_factors = tuple(s / i for s, i in zip(spatial_size, inflated_spatial_size))
+    scale_mat = factory.scale(scale_factors)
+    return matmul(scale_mat, matrix)
+
+
+class Matrix:
+    def __init__(self, matrix: NdarrayOrTensor):
+        self.data = ensure_tensor(matrix)
+
+    # def __matmul__(self, other):
+    #     if isinstance(other, Matrix):
+    #         other_matrix = other.data
+    #     else:
+    #         other_matrix = other
+    #     return self.data @ other_matrix
+    #
+    # def __rmatmul__(self, other):
+    #     return other.__matmul__(self.data)
+
+
+class Grid:
+    def __init__(self, grid):
+        self.data = ensure_tensor(grid)
+
+    # def __matmul__(self, other):
+    #     raise NotImplementedError()
+
+
+class MetaMatrix:
+    def __init__(self, matrix: Union[NdarrayOrTensor, Matrix, Grid], metadata: Optional[dict] = None):
+
+        if not isinstance(matrix, (Matrix, Grid)):
+            if matrix.shape == 2:
+                if matrix.shape[0] != matrix.shape[1] or matrix.shape[0] not in (3, 4):
+                    raise ValueError(
+                        "If 'matrix' is passed a numpy ndarray/torch Tensor, it must"
+                        f" be 3x3 or 4x4 ('matrix' has has shape {matrix.shape})"
+                    )
+            matrix_ = Matrix(matrix)
+        else:
+            matrix_ = matrix
+        self.matrix = matrix_
+
+        self.metadata = metadata or {}
+
+    def __matmul__(self, other):
+        if isinstance(other, MetaMatrix):
+            other_ = other.matrix
+        else:
+            other_ = other
+        return MetaMatrix(self.matrix @ other_)
+
+    def __rmatmul__(self, other):
+        if isinstance(other, MetaMatrix):
+            other_ = other.matrix
+        else:
+            other_ = other
+        return MetaMatrix(other_ @ self.matrix)
+
+
+def matmul(
```

Comment:
this util almost addresses https://github.com/Project-MONAI/MONAI/issues/4604 if the interface is simpler and considers a batch of transforms... any interests in tackling that ticket within this implementation? @atbenmurray

Reply:
@wyli sorry, access has been patchy. Let me take a look

Reply:
@wyli Yes, I think so. I haven't understood yet how to do it so if there are well defined methods that is great!

Reply:
@wyli Hmm, it does appear to be the same principle. Well, at least we are pytorch and they are tensorflow. I did look across a number of libraries but I focused on the pytorch-based ones of course

Reply:
@wyli hmm, it seems to be more focused on providing these capabilities for registration based deep learning, as opposed to preprocessing

Reply:
if this `matmul` could be the util to implement the key idea and shared between preprocessing and as a differentiable module , that would be nice...